src/commonMain/kotlin/arrow.exact/Exact.kt

```diff
@@ -1,23 +1,33 @@
 package arrow.exact
 
-import arrow.core.*
+import arrow.core.Either
 
-interface Exact<A, out B> {
 
-  fun from(value: A): Either<ExactError, B>
+interface Exact<A, out R : Refined<A>> : ExactEither<ExactError, A, R>
 
-  fun fromOrNull(value: A): B? {
-    return from(value).getOrNull()
-  }
+data class ExactError(val message: String)
+
+interface ExactEither<out E : Any, A, out R : Refined<A>> {
+
+  fun from(value: A): Either<E, R>
+
+  // TODO: This doesn't work for some weird reason :/
+  // fun Raise<E>.fromOrRaise(value: A): R = from(value).bind()
 
-  fun fromOrThrow(value: A): B {
-    return when (val result = from(value)) {
-      is Either.Left -> throw ExactException(result.value.message)
-      is Either.Right -> result.value
-    }
+  fun fromOrNull(value: A): R? = from(value).getOrNull()
+
+  fun fromOrThrow(value: A): R = when (val result = from(value)) {
+    is Either.Left -> throw ExactException(result.value)
+    is Either.Right -> result.value
   }
+
+  // TODO: What are your thoughts about this?
+  operator fun invoke(value: A): R = fromOrThrow(value)
```

Comment:
This is a bit controversial. What do you think?
My opinion is that there are many cases where you're sure that something matches the constraints and going through the hassle of dealing with an `Either` is unnecessary.
```kotlin
// here I'm certain that positive + positive is always a positive
fun plus(x: PositiveInt, y: PositiveInt): PositiveInt = PositiveInt(x.value + y.value)
```

Reply:
I personally think `orThrow` is concise enough it's worth the readability. Kotlin favours explicitness in a lot of areas, and this seems a place that is a good trade-off for it as well.

Reply:
Yes, that's true @nomisRev . IMO, it just becomes too verbose, distracting and not right making the possibility of throwing an exception for cases where that's obviously impossible like:
```kotlin
val hello = NotBlankTrimmedString.fromOrThrow("Hello")
val world = NotBlankTrimmedString.fromOrThrow("wold!")
val helloWorld = NotBlankTrimmedString.fromOrThrow(hello.value + " " + world.value)
```
Personal opinion, the above is too much noise and I prefer the below:
```kotlin
val hello = NotBlankTrimmedString("Hello")
val world = NotBlankTrimmedString("wold!")
val helloWorld = NotBlankTrimmedString(hello.value + " " + world.value)
```
Note these two are artificial examples but in practice we have many places where we're certain that a constraint won't be violated, in my personal projects I reserve the `operator fun invoke(value: A): R = fromOrThrow(value)` for such cases. 

When using for example `PositiveInt(x + y)` instead of `PositiveInt.fromOrThrow(x + y)` we show our readers that we're certain that in the first case `x + y` will always be positive but in the 2nd we direct their attention to knowing that it's not always the case.

I suggest leaving it for discussion and re-visiting before releasing the library.